File: Leetcode/Medium/309. Best Time to Buy and Sell Stock with Cooldown/309. Best Time to Buy and Sell Stock with Cooldown.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function (prices) {
    let hold, sold, rest;
    hold = [Number.MIN_SAFE_INTEGER, ...Array(prices.length).fill(0)];
    sold = Array(prices.length + 1).fill(0);
    rest = Array(prices.length + 1).fill(0);

    for (let i = 1; i < prices.length + 1; ++i) {
        hold[i] = Math.max(hold[i - 1], rest[i - 1] - prices[i - 1]);
        sold[i] = hold[i - 1] + prices[i - 1];
        rest[i] = Math.max(rest[i - 1], sold[i - 1]);
    }

    return Math.max(sold[prices.length], rest[prices.length]);
};

console.log(
    maxProfit([1, 2, 3, 0, 2])
)

File: Leetcode/Medium/309. Best Time to Buy and Sell Stock with Cooldown/309. Best Time to Buy and Sell Stock with Cooldown.py
from typing import List
import sys


class Solution:
    def maxProfit(self, prices: List[int]) -> int:

        hold = [0] * (len(prices) + 1)
        hold[0] = -sys.maxsize - 1
        sold = [0] * (len(prices) + 1)
        rest = [0] * (len(prices) + 1)

        for i in range(len(prices)):
            hold[i + 1] = max(hold[i], rest[i] - prices[i])
            sold[i + 1] = hold[i] + prices[i]
            rest[i + 1] = max(rest[i], sold[i])
            pass

        return max(sold[len(prices)], rest[len(prices)])


d = Solution()
d.maxProfit([1, 2, 3, 0, 2])
